### ISSUE TITLE
Availability State Missing Features

### ISSUE THREAD

**Issue by shangyian** [DataJunction/dj#691]
Let’s say we want to schedule materialization for a transform node that has several upstream transforms.

We then need to determine (a) if those transforms can be built all the way down to source nodes or (b) if they MUST be materialized (i.e., uses DJ_LOGICAL_TIMESTAMP()).

If (a), we can build the node’s query all the way down and schedule a materialization job with that query. However, any time an upstream node’s transform has an availability state posted, we will need to update the downstream node’s materialization job to use the transform.

If (b), we have two options:
1. We still schedule the node’s materialization but we’ll assume that the upstream transform will be materialized to a particular table name. Then the scheduled job will just wait for the to-be-materialized upstream table to be ready before starting.
2. Alternatively we can block on scheduling this node’s materialization until the upstream transform is materialized.

There are pros and cons to both approaches. In approach 2, we force users to acknowledge that there is an expected dependencies DAG with materialization, and that they need to successfully materialize some nodes before others can be materialized. In approach 1, we’re more flexible and let users schedule at their own pace. However, users who don’t understand the dependencies DAG may be confused. (@agorajek I think we discussed this before at some point and you preferred 1?)

**Comment by agorajek** [comment]
@shangyian there is a lot to unpack in your post, but let me start with a short call that I think we should implement both (a) and (b1) together. I don't think (b2) is useful because it means tell the user: "come back tomorrow and maybe we can schedule your materialization". I feel like this is not a good experience.

So let me clarify what I mean by (a) and (b1) together, and for a moment let's focus on the latest/incremental load. So if someone makes a transform node and schedules a materialization, we should:
1. other than the obvious: setting the materlialization config and some conveniance flags at the node level
2. kick of a scheduler job create/update process, which should involve:
     - (i)looking at all upstream nodes and their current materialization situations
     - (ii) designing an optimal job for the latest partition, and by optimal I mean: using the closest materialized parents
     - (iii) schedule this job and finish.
3. next day (or hour, depending on the schedule) the materialization service, should scan through all the nodes with a materialization configs and redo all the steps above (i - iii). 

This way we can be sure that the materialization jobs will always adjust to the new configuration of the system, barring some race conditions, which we can monitor for. Does that make sense @shangyian ?

The thing that is harder to implement are **backfills** because the upstream nodes could be materialized for different periods and not have a continuous range that can be used. I still think that this problem could be solved by a Backfill Design Helper page, where we help the user make the right call and raise errors if an impossible backfill is generated. But we can make little steps in this feature space.

**Comment by shangyian** [comment]
@agorajek What you described all makes sense to me! 

> designing an optimal job for the latest partition, and by optimal I mean: using the closest materialized parents

This point here actually makes me realize the importance of #688, because the bottleneck around the AST is not just at the metric building stage, but also comes up when we want to materialize nodes. In the scenario described above, every time we want to design the optimal job for the latest partition, we'll need to reparse, recompile, and rebuild the node's query in order to generate the right materialization query. That means that maybe stage 0 to this whole thing is actually getting AST serialization in place, which shouldn't take very long imo.

However, another scenario I'm thinking about is this: when we reference an upstream in a node's query, we don't specify what partitions we expect to match for, which makes determining whether the upstream's already materialized table partitions satisfy the current node's needs. 

Here's an example -- let's say we have the following nodes:
```
node: A
query: select one, two, dj_logical_timestamp() as date_partition from B
partition columns: date_partition

node: B
query: select dateint, three as one, four as two from C
partition columns: dateint
```

Node B doesn't have `dj_logical_timestamp()` and its partition columns don't match A. Does this just mean that any materialized table for B will work for A? How would we know that the `dateint` partition is meant to match the `date_partition` in node A, unless we just assume that if they're both temporal columns of the same unit (date), they'll match?

**Comment by agorajek** [comment]
@shangyian:
> In the scenario described above, every time we want to design the optimal job for the latest partition, we'll need to reparse, recompile, and rebuild the node's query in order to generate the right materialization query

Yes, that's correct and making this step more performant would be great. But remember that this is mostly in the batch mode space, so the user would not necessarily need to wait for this if it had to take minutes.

> Here's an example -- let's say we have the following nodes... unless we just assume that if they're both temporal columns of the same unit (date), they'll match?

I like this example, and I think it is one of the cases which makes an argument for a good design around temporary partitions. DJ simply needs to "understand" the temporal partitions and the time grains to make these choices.
